x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -97,7 +97,45 @@ func (k msgServer) RelayPayment(goCtx context.Context, msg *types.MsgRelayPaymen
 				)
 			}
 
-			// badge is valid -> switch address to badge signer (developer key) and continue with payment
+			badgeUsedCuMapKey := k.CreateBadgeUsedCuMapKey(badgeData.Badge.ProjectSig, relay.Provider)
+			badgeUsedCuMapEntry, found := k.GetBadgeUsedCu(ctx, badgeUsedCuMapKey)
+			if found {
+				if relay.CuSum <= badgeData.Badge.CuAllocation-badgeUsedCuMapEntry.UsedCu {
+					badgeUsedCuMapEntry.UsedCu += relay.CuSum
+					k.SetBadgeUsedCu(ctx, badgeUsedCuMapEntry)
+				} else {
+					return nil, utils.LavaFormatWarning("badge CU allocation exceeded", fmt.Errorf("could not update badge's used CU"),
+						utils.Attribute{Key: "relayCuSum", Value: relay.CuSum},
+						utils.Attribute{Key: "badgeCuLeft", Value: badgeData.Badge.CuAllocation - badgeUsedCuMapEntry.UsedCu},
+					)
+				}
+			} else {
+				if relay.CuSum <= badgeData.Badge.CuAllocation {
+					// setting a timer with a callback to delete the badgeUsedCuEntry after epochBlocks*blocksToSave (see keeper.go)
+					badgeUsedCuTimerExpiryBlock := k.BadgeUsedCuExpiry(ctx, badgeData.Badge)
+					if badgeUsedCuTimerExpiryBlock <= uint64(ctx.BlockHeight()) {
+						return nil, utils.LavaFormatWarning("badge used CU entry expired", fmt.Errorf("could not create badge used CU entry"),
+							utils.Attribute{Key: "badgeUsedCuTimerExpiryBlock", Value: badgeUsedCuTimerExpiryBlock},
+							utils.Attribute{Key: "currentBlock", Value: uint64(ctx.BlockHeight())},
+						)
+					}
+					// timerKey = badgeUsedCuMapKey since all badgeUsedCuMapKey keys are unique - can be used to differentiate the timers
+					k.badgeTimerStore.AddTimerByBlockHeight(ctx, badgeUsedCuTimerExpiryBlock, badgeUsedCuMapKey, badgeUsedCuMapKey)
+
+					badgeUsedCuMapEntry = types.BadgeUsedCu{
+						BadgeUsedCuMapKey: badgeUsedCuMapKey,
+						UsedCu:            relay.CuSum,
+					}
+					k.SetBadgeUsedCu(ctx, badgeUsedCuMapEntry)
+				} else {
+					return nil, utils.LavaFormatWarning("badge CU allocation exceeded", fmt.Errorf("could not create badge used CU entry"),
```

Comment:
how about the following messages instead:
```"badge rejected", fmt.Errorf("badge CU allocation exceeded")```
and above could be
```"badge rejected", fmt.Errorf("badge validity expired")```
etc.
I assume this goes back to the user, and this is more informative and less implementation dependent.

Reply:
done